Ndev 2478 add send erc20 scenario

romanova-natasha opened the pull request:
add send erc20 scenario
add metrics for erc20 scenario
update readme for k6 load tests, minor refactoring

romanova-natasha committed:
add contract deployment

add contract deployment

romanova-natasha committed:
add erc20 metrics, add setup step, deploy contract before test

add erc20 metrics, add setup step, deploy contract before test

romanova-natasha committed:
update grafana dashboard: add erc20 params

update grafana dashboard: add erc20 params

romanova-natasha committed:
get gas limit and gas price in newContract method inside xk6 plugin

get gas limit and gas price in newContract method inside xk6 plugin

romanova-natasha committed:
add: build transfer erc20 tx and execute it

add: build transfer erc20 tx and execute it

romanova-natasha committed:
set signer initial balance

set signer initial balance

romanova-natasha committed:
change metrics dashboards: fix legends, add all params in aggregated metrics

change metrics dashboards: fix legends, add all params in aggregated metrics

romanova-natasha committed:
make separate commands for different steps before load test scenario run, update readme

make separate commands for different steps before load test scenario run, update readme

romanova-natasha committed:
Merge branch 'develop' into NDEV-2478-add-send-erc20-scenario

kristinaNikolaevaa commented on clickfile.py:
```diff
@@ -1265,18 +1265,45 @@ def build(tag):
+@click.option("-a", "--bank_account", default=None, required=False, help="Bank account address if needed")
 @catch_traceback
-def run(network, script, users, balance, bank_account):
+def deploy_erc20(network, balance, bank_account):
```
are you sure that we need all this commands to prepare data? 
I suggest to have one command for "run load tests" which will include all needed preparation  (contracts deploy, account creating e.t.c)

romanova-natasha replied:
done

kristinaNikolaevaa commented on clickfile.py:
```diff
@@ -1265,18 +1265,45 @@ def build(tag):
+
+@k6.command("run", help="Run k6 performance test.")
+@click.option("-n", "--network",required=True, default="local", help="Which network to use for envs assignment")
+@click.option("-s", "--script", required=True, default="./loadtesting/k6/tests/sendNeon.test.js", help="Path to k6 script")
```
Lets make default value = to run all possible load tests for proxy ?

romanova-natasha replied:
I've planned to collect all scenarios for proxy in a common one. Then run it taking into account weight of each separate scenario. Do you want to have it in this branch?

kristinaNikolaevaa replied:
ok, let's do it later

kristinaNikolaevaa commented on loadtesting/k6/readme.md:
````diff
@@ -7,6 +7,11 @@ Make sure you have Golang bin path in PATH, otherwise
 ```
 
+### Compile contracts
+Make sure you have solc installed and run to generate an abi file for erc20 contract:
````
can we do it inside the script?

romanova-natasha replied:
done

kristinaNikolaevaa commented on loadtesting/k6/readme.md:
````diff
@@ -28,11 +33,27 @@ Install xk6 and build an executable file, tag is a neonlabsorg forked xk6-ethere
 ```
 
-It is mandatory either set up `K6_USERS_NUMBER` and `K6_INITIAL_BALANCE` envs or pass `--users` and `--balance` flags to run the command. We need these values to prepare test accounts for further transfers transactions. 
+It is mandatory either set up `K6_USERS_NUMBER` and `K6_INITIAL_BALANCE` envs or pass `--users` and `--balance` flags to run the command. We need these values to set up options for k6 load scenario.
````
As I understand you removed the possibility to set K6_USERS_NUMBER and K6_INITIAL_BALANCE
Only option is to pass `--users` and `--balance`

romanova-natasha replied:
fixed

kristinaNikolaevaa commented on loadtesting/k6/readme.md:
````diff
@@ -28,11 +33,27 @@ Install xk6 and build an executable file, tag is a neonlabsorg forked xk6-ethere
 ```
 
-It is mandatory either set up `K6_USERS_NUMBER` and `K6_INITIAL_BALANCE` envs or pass `--users` and `--balance` flags to run the command. We need these values to prepare test accounts for further transfers transactions. 
+It is mandatory either set up `K6_USERS_NUMBER` and `K6_INITIAL_BALANCE` envs or pass `--users` and `--balance` flags to run the command. We need these values to set up options for k6 load scenario.
````
How can I understand how big the balance should be? 
Maybe it's better to count the value inside and remove the flag ?

romanova-natasha replied:
we will add initial balance calculation/additional funding during the scenario run later

kristinaNikolaevaa commented on utils/k6_helpers.py:
```diff
@@ -28,10 +25,30 @@ def k6_prepare_accounts(network, network_object, users, balance, bank_account):
         json.dump(accounts, f)
 
+def deploy_erc20_contract(web3_client, faucet, account):
+    erc20 = ERC20(
```
wrong ERC20 ...
It is better to check erc20ForSpl

romanova-natasha replied:
we will add erc20ForSpl scenario later

kristinaNikolaevaa commented on loadtesting/k6/tests/sendErc20.test.js:
```diff
@@ -0,0 +1,49 @@
+const abi = JSON.parse(open(pathToContractData + '.abi'));
+
+export default function sendErc20Test() {
+    // We need to have a unique signer account for each VU
```
I think by this way we will get not real scenario
You can do transferring some tokens to all accounts as part of the preparation step 
and then sent tokens from different senders here

romanova-natasha replied:
fixed

kristinaNikolaevaa commented on loadtesting/k6/tests/utils/ethClient.js:
```diff
@@ -12,11 +12,16 @@ export function ethClient(privateKey) {
+export function sendErc20ViaTransferFunction(client, erc20, owner, abi, signerAddress, receiverAddress, amount) {
+    const input = erc20.fillInput(abi, "transfer");
+    const nonce = client.getNonce(signerAddress);
+    return sendTokens(client, owner, receiverAddress, amount, input, null, null, nonce)
```
looks strange 
for erc20 "amount" should be part of input field (or data field) 
and this ""value": value," should be 0 all the time

romanova-natasha replied:
fixed: full changes of input calculations are in xk6 plugin

kristinaNikolaevaa commented on loadtesting/k6/tests/utils/accountManager.js:
```diff
@@ -0,0 +1,63 @@
+    };
+}
+
+export function createFundedAccount(client, amount, bankAccount) {
```
do you use this function somewhere?

romanova-natasha replied:
it is removed now

romanova-natasha committed:
set predefined accounts for send erc20 scenario, fix review notes, add receipt status error metrics

set predefined accounts for send erc20 scenario, fix review notes, add receipt status error metrics

kristinaNikolaevaa commented on loadtesting/k6/tests/sendErc20.test.js:
```diff
@@ -0,0 +1,65 @@
+            client,
+            erc20,
+            accountSenderAddress,
+            erc20Address,
```
does erc20 object contain its address?

romanova-natasha replied:
No, it is a private field. I removed erc20Address from functions args, I agree it is a bit overloaded here

kristinaNikolaevaa commented on loadtesting/k6/tests/sendErc20.test.js:
```diff
@@ -0,0 +1,65 @@
+            erc20,
+            accountSenderAddress,
+            erc20Address,
+            JSON.stringify(abi),
```
you can define abi inside the fuction because it can't be different for this function

romanova-natasha replied:
We use open() function to read the .abi file. This function can be used only at the init stage. Unfortunately we can't move it to ethClient.js file and we can't use smth similar cause we do not have nodejs here

romanova-natasha committed:
fix after the review

fix after the review
